src/DLPack.jl

```diff
@@ -149,63 +191,90 @@ dtypes_to_jltypes() = Dict(
+device_type(tensor::DLManagedTensor) = device_type(tensor.dl_tensor)
+device_type(manager::DLManager) = device_type(manager.tensor)
 
 Base.eltype(array::DLArray{T}) where {T} = T
 
 Base.ndims(array::DLArray{T, N}) where {T, N} = N
 
-_size(tensor::DLTensor) = tensor.shape
-_size(manager::DLManagedTensor) = _size(manager.dl_tensor)
-_size(array::DLArray) = _size(array.manager)
-
-function Base.size(array::DLArray{T, N}) where {T, N}
-    ptr = Base.unsafe_convert(Ptr{NTuple{N, Int64}}, _size(array))
-    return unsafe_load(ptr)
-end
+unsafe_size(manager::DLManager{T, N}) where {T, N} = unsafe_size(manager.tensor, Val(N))
 #
-function Base.size(array::DLArray{T, N}, d::Integer) where {T, N}
-    if 1 ≤ d
-        return d ≤ N ? size(array)[d] : Int64(1)
-    end
-    throw(ArgumentError("Dimension out of range"))
+function unsafe_size(tensor::DLManagedTensor, ::Val{N}) where {N}
+    sz = tensor.dl_tensor.shape
+    ptr = Base.unsafe_convert(Ptr{NTuple{N, Int64}}, sz)
+    return unsafe_load(ptr)
 end
 
-_strides(tensor::DLTensor) = tensor.strides
-_strides(manager::DLManagedTensor) = _strides(manager.dl_tensor)
-_strides(array::DLArray) = _strides(array.manager)
+Base.size(array::DLArray) = size(array.data)
+Base.size(array::DLArray, d::Integer) = size(array.data, d)
 
-function Base.strides(array::DLArray{T, N}) where {T, N}
-    ptr = Base.unsafe_convert(Ptr{NTuple{N, Int64}}, _strides(array))
+function unsafe_strides(tensor::DLManagedTensor, val::Val{N}) where {N}
+    st = tensor.dl_tensor.strides
+    if st == C_NULL
+        trailing_size = Base.rest(unsafe_size(tensor, val), 2)
+        tup = ((reverse ∘ cumprod ∘ reverse)(trailing_size)..., 1)
+        return NTuple{N, Int64}(tup)
+    end
+    ptr = Base.unsafe_convert(Ptr{NTuple{N, Int64}}, st)
     return unsafe_load(ptr)
 end
 
+Base.strides(a::DLArray) = strides(a.data)
+
 byte_offset(tensor::DLTensor) = Int(tensor.byte_offset)
-byte_offset(manager::DLManagedTensor) = byte_offset(manager.dl_tensor)
-byte_offset(array::DLArray) = byte_offset(array.manager)
+byte_offset(tensor::DLManagedTensor) = byte_offset(tensor.dl_tensor)
+byte_offset(manager::DLManager) = byte_offset(manager.tensor)
 
 Base.pointer(tensor::DLTensor) = tensor.data
-Base.pointer(manager::DLManagedTensor) = pointer(manager.dl_tensor)
-Base.pointer(array::DLArray) = pointer(array.manager)
-
-function Base.unsafe_wrap(::Type{Array}, array::DLArray{T}) where {T}
-    if device_type(array) == kDLCPU
-        addr = Int(pointer(array))
-        return GC.@preserve array unsafe_wrap(Array, Ptr{T}(addr), size(array))
+Base.pointer(tensor::DLManagedTensor) = pointer(tensor.dl_tensor)
+Base.pointer(manager::DLManager) = pointer(manager.tensor)
+
+function Base.unsafe_wrap(::Type{Array}, manager::DLManager{T}) where {T}
+    if device_type(manager) == kDLCPU
+        addr = Int(pointer(manager))
+        sz = unsafe_size(manager)
+        return GC.@preserve manager unsafe_wrap(Array, Ptr{T}(addr), sz)
```

Comment:
This doesn't take ownership of the pointer, right? Overall I'm still confused who handles the GC for DLPack managed array. In Python land everything is python object so only has to interact with a single GC, but here things are a little more confusing.

Reply:
This is how PyCall.jl handles numpy arrays: https://github.com/JuliaPy/PyCall.jl/blob/master/src/pyarray.jl#L326. Notice it doesn't take ownership and there is no `GC.@preserve`.

Reply:
The default for `unsafe_wrap` is `own = false` so this doesn't take ownership, the `GC.@preserve` here is just an extra precaution.

The thing is that the `PyCapsule` that `to_dlpack` generates (and not the original array, that is a separate object) is supposed to pass the responsibility of freeing the memory of the `DLManagedTensor` to however is importing, but julia's garbage collector doesn't know that the array generated is tied to the lifetime of the `DLManagedTensor` so it can sometimes think it's no longer needed (unless the `DLArray` lives on global scope or it's `GC.@preserve`d).

It's going to be extremely hard to make this safe for the user. What I can do instead is document that whenever you are using a `DLArray` (that doesn't live in global scope) you will need to do it like this:

```julia
dla = DLArray(tensor, to_dlpack)

GC.@preserve dla begin
    # do whatever you want with dla.data
end
```

Reply:
Let me think about this a bit more, For now, I think we can go ahead and merge this.

Reply:
Actually, I think that after this PR, you should no longer see the issue in #14 even without the `CG.@preserve` I wrote above.